### PR TITLE
Simplify lint() internals towards cyclomatic complexity target

### PR DESCRIPTION
**Pull request by MichaelChirico**
Towards #653

This pass reduced us from 28-->19. I think that's pretty good for now.

This enables us to tighten our belt on the lintr down to 20 hereafter.

### CODE REVIEW

File: R/lint.R [hunk 1]
```diff
@@ -100,32 +90,17 @@ lint <- function(filename, linters = NULL, cache = FALSE, ..., parse_settings =
...
-        }
-      }
-      lints[[itr <- itr + 1L]] <- expr_lints
-    }
-  }
+  n_linters <- length(linters)
+  n_checks <- length(source_expressions$expressions) * n_linters
+  lints <- lapply(seq_len(n_checks), function(check_i) {
+    # modular arithmetic to vary linters first, then expressions
+    expr <- source_expressions$expressions[[((check_i - 1L) %/% n_linters) + 1L]]
```
Comment (MichaelChirico) [inline R/lint.R:97]: not the prettiest -- two other approaches:

 1. Use `DF <- expand.grid(seq_len(n_linters), seq_len(n_expr))`, then `DF[check_i, ]` has the right coords
 2. Use `M <- matrix(nrow = n_linters, ncol = n_expr)`, then `col(M)` and `row(M)` have the right coords

Comment (AshesITR) [reply R/lint.R:97]: I think we're sacrificing readability for cyclomatic complexity here.
NB: did you take a look at @russHyde's PR for caching as a class?

Comment (MichaelChirico) [reply R/lint.R:97]: Looks like a more systematic way of accomplishing what's attempted in this PR. Nice! Any hope for progress @russHyde ? :grimacing:

**Comment by MichaelChirico** [comment]
Closing for now, but leaving the branch in case there's anything salvageable.

**Pull request closed without merging**